### PR TITLE
Add specific 'defined?(@ivar)' instruction

### PR DESCRIPTION
**Pull request by olefriis**
Currently, the `defined?` operation is implemented with single, general VM instruction covering different cases of what's being checked. This PR creates a specific VM instruction for when the argument to `defined?` is an instance variable.

This allows us to eliminate a `putnil` instruction for that case, and the implementation is more straightforward. Additionally, we can make use of shape information in CRuby.

I wrote a very silly benchmark:

```ruby
require 'benchmark'

class IVarClass
  def set_ivar
    @foo = 1
  end

  def test_defined
    50000000.times { defined?(@foo) }
  end
end

obj = IVarClass.new
puts Benchmark.measure { obj.test_defined }

obj.set_ivar
puts Benchmark.measure { obj.test_defined }
```

I've run it a few times on `master` and this branch, with and without YJIT enabled. (I'm on a laptop, so runtimes may fluctuate a bit, so this is not exact science.)

Representative values on `master`:
```
$ ./ruby --yjit ivar_microbenchmark.rb
  3.580647   0.006810   3.587457 (  3.590800)
  3.516924   0.004903   3.521827 (  3.525773)
$ ./ruby ivar_microbenchmark.rb
  4.016235   0.008735   4.024970 (  4.029170)
  4.032165   0.006480   4.038645 (  4.041866)
```

On this branch:
```
$ ./ruby --yjit ivar_microbenchmark.rb
  3.172839   0.006284   3.179123 (  3.182953)
  3.125201   0.006086   3.131287 (  3.133984)
$ ./ruby ivar_microbenchmark.rb 
  3.447063   0.007094   3.454157 (  3.457770)
  3.490655   0.007111   3.497766 (  3.501535)
```

For these runs, it gives about 13% better performance on CRuby and about 11% better performance with YJIT. As mentioned, these numbers are not super stable on my laptop, but I'm consistently getting better than 10% both on CRuby and YJIT.

## Notes
For the CRuby version, I've added an extra parameter to `vm_getivar` so that I can re-use the logic. I'm not super proud of this, as it makes an already complex function a bit more complex. I've tried to be as gentle as possible. If somebody has other ideas to get the same effect in a cleaner way, let me know.

Also, this _will_ introduce a new internal instruction. I don't know how people feel about that.

Thanks a lot to @jhawthorn who has helped from start to finish with this PR!

### CODE REVIEW

**Review by maximecb — APPROVED** [review]

**Comment by maximecb** [comment]
I feel like this is mostly an optimization for the interpreter. No reason the JIT can't use shapes to optimize `defined` already, it just hasn't been done. However, not strongly opposed to this. It's not a bad thing to make the interpreter faster, so I approved it.

**Comment by k0kubun** [comment]
How much impact did it have on non-microbenchmarks?

**Comment by tenderlove** [comment]
I think this is a very good idea for the interpreter.  I'm not a fan of people doing `@foo = blah unless defined?(@foo)`, but it's something people do in the wild and we may as well have an IC for that case.

**Comment by maximecb** [comment]
> I think this is a very good idea for the interpreter. I'm not a fan of people doing `@foo = blah unless defined?(@foo)`, but it's something people do in the wild and we may as well have an IC for that case.

Good point Aaron 👍

**Comment by k0kubun** [comment]
[yjit-bench](https://github.com/shopify/yjit-bench) `--category headline` results:

## Interpreter
```
before: ruby 3.3.0dev (2023-02-27T11:07:19Z master 0aa50a03b1) [x86_64-linux]
after: ruby 3.3.0dev (2023-03-02T12:07:49Z add-defined_ivar-i.. 55a5156fc0) [x86_64-linux]

-------------  -----------  ----------  ----------  ----------  ------------  -------------
bench          before (ms)  stddev (%)  after (ms)  stddev (%)  before/after  after 1st itr
activerecord   114.7        0.3         112.2       0.6         1.02          1.01
hexapdf        1809.9       0.7         1756.3      0.7         1.03          1.02
liquid-c       44.4         1.9         44.7        1.9         0.99          0.99
liquid-render  109.3        1.3         109.6       1.5         1.00          0.99
mail           94.8         0.3         93.3        0.6         1.02          1.00
psych-load     1453.5       0.2         1422.2      0.1         1.02          1.02
railsbench     1530.3       0.9         1522.3      1.0         1.01          1.01
ruby-lsp       47.1         12.8        46.3        13.1        1.02          0.99
sequel         90.1         1.8         90.4        1.7         1.00          1.00
-------------  -----------  ----------  ----------  ----------  ------------  -------------
```

## YJIT
```
before: ruby 3.3.0dev (2023-02-27T11:07:19Z master 0aa50a03b1) +YJIT [x86_64-linux]
after: ruby 3.3.0dev (2023-03-02T12:07:49Z add-defined_ivar-i.. 55a5156fc0) +YJIT [x86_64-linux]

-------------  -----------  ----------  ----------  ----------  ------------  -------------
bench          before (ms)  stddev (%)  after (ms)  stddev (%)  before/after  after 1st itr
activerecord   67.4         0.5         67.5        0.3         1.00          1.00
hexapdf        1213.3       1.2         1204.9      1.4         1.01          1.00
liquid-c       32.0         3.2         32.2        2.9         0.99          1.00
liquid-render  62.5         2.9         62.4        3.3         1.00          0.99
mail           72.7         0.2         72.1        0.3         1.01          1.00
psych-load     1022.7       0.2         1014.9      0.1         1.01          1.00
railsbench     1004.8       1.3         1004.2      1.7         1.00          1.00
ruby-lsp       37.0         28.0        36.7        29.6        1.01          1.01
sequel         80.1         2.5         80.1        2.3         1.00          1.01
-------------  -----------  ----------  ----------  ----------  ------------  -------------
```

---

The interpreter seems 2~3% faster on some benchmarks (I can stably reproduce it). YJIT seems to be the same.

**Comment by maximecb** [comment]
@jhawthorn @olefriis Aaron @tenderlove has shown interest in implementing a faster version of this instruction for YJIT which takes advantage of shapes.

Would you be willing to pair with him on this? :)

File: insns.def
Comment (ko1) [inline insns.def:706]: YARV instruction name has a rule "do not contain `_` other than optimization instructions".
Can I rename it?

Comment (olefriis) [reply insns.def:706]: I would consider this an optimization instruction? But apart from that, totally fine with me renaming it.

(...though I'll defer to more experienced committers to decide here...)

Comment (ko1) [reply insns.def:706]: > I would consider this an optimization instruction?

with compile option optimized instructions should be disable.
anyway I'll rearrange it.
Thank you.

**Pull request merged**
